notifications/manager.go

```diff
@@ -0,0 +1,210 @@
+package notifications
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/lightninglabs/loop/swapserverrpc"
+	"google.golang.org/grpc"
+)
+
+// NotificationType is the type of notification that the manager can handle.
+type NotificationType int
+
+const (
+	// NotificationTypeUnknown is the default notification type.
+	NotificationTypeUnknown NotificationType = iota
+
+	// NotificationTypeReservation is the notification type for reservation
+	// notifications.
+	NotificationTypeReservation
+)
+
+// Client is the interface that the notification manager needs to implement in
+// order to be able to subscribe to notifications.
+type Client interface {
+	// SubscribeNotifications subscribes to the notifications from the server.
+	SubscribeNotifications(ctx context.Context,
+		in *swapserverrpc.SubscribeNotificationsRequest,
+		opts ...grpc.CallOption) (
+		swapserverrpc.SwapServer_SubscribeNotificationsClient, error)
+}
+
+// Config contains all the services that the notification manager needs to
+// operate.
+type Config struct {
+	// Client is the client used to communicate with the swap server.
+	Client Client
+
+	// FetchL402 is the function used to fetch the l402 token.
+	FetchL402 func(context.Context) error
+}
+
+// Manager is a manager for notifications that the swap server sends to the
+// client.
+type Manager struct {
+	cfg *Config
+
+	hasL402 bool
+
+	subscribers map[NotificationType][]subscriber
+	sync.Mutex
+}
+
+// NewManager creates a new notification manager.
+func NewManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:         cfg,
+		subscribers: make(map[NotificationType][]subscriber),
+	}
+}
+
+type subscriber struct {
+	subCtx   context.Context
+	recvChan interface{}
+}
+
+// SubscribeReservations subscribes to the reservation notifications.
+func (m *Manager) SubscribeReservations(ctx context.Context,
+) <-chan *swapserverrpc.ServerReservationNotification {
+
+	notifChan := make(chan *swapserverrpc.ServerReservationNotification, 1)
+	sub := subscriber{
+		subCtx:   ctx,
+		recvChan: notifChan,
+	}
+
+	m.Lock()
+	m.subscribers[NotificationTypeReservation] = append(
```

Comment:
nit:  consider also extracting this as a method so we have `addSubscriber` and `removeSubscriber` for when we add more subscription types.